nntrainer/dataset/data_producers.h

```diff
@@ -77,25 +97,63 @@ class DataProducer {
 
   /**
    * @brief finalize the class with given properties
+   * @todo remove this
    * @return Generator generator is a function that generates an iteration upon
    * call
    *
    */
-  virtual Generator finalize(const std::vector<TensorDim> &input_dims,
-                             const std::vector<TensorDim> &label_dims) = 0;
+  [[deprecated("use finalize_sample instead")]] virtual Generator
+  finalize(const std::vector<TensorDim> &input_dims,
+           const std::vector<TensorDim> &label_dims) = 0;
+
+  /**
+   * @brief finalize the class to return a immutable Generator.
+   * @todo rename this to finalize.
+   * @remark this function must assume that the batch dimension of each tensor
+   * dimension is one. If actual dimension is not one, this function must ignore
+   * the batch dimension and assume it to be one.
+   * @param input_dims input dimensions.
+   * @param label_dims label dimensions.
+   * @return Generator generator is a function taht generates a sample upon
+   * call.
+   */
+  virtual Generator_sample
+  finalize_sample(const std::vector<TensorDim> &input_dims,
```

Comment:
i recommend to name this `finalize`, and rename deprecated version to `finalize_batch`.

Reply:
`finalize_batch` will be deleted within two days, I'll keep this to not confuse the reviewers (from previous commits that is not merged) but rename it at the end of the work :)

Reply:
alright.